File: src/export_anim_numbers.py
"""Export the real numbers the Manim animation shows, and prove they match the analysis.

For DNAJB7 (EUR87_r1 vs YRI87) this computes log V for all 8 mixes of weights, allele
frequencies and LD, recomputes the three Shapley components from those 8 values, and
checks them against results/decomposition/EUR87_r1_vs_YRI87.tsv. The animation reads only
the JSON written here, so it cannot show a number the analysis did not produce.
Uses the dosage exports written by src/fig12_locus.py.
"""
import itertools
import json
import math
from pathlib import Path

import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
LOC = ROOT / "results" / "figdata" / "locus"
GENE, NAME = "ENSG00000172404.4", "DNAJB7"

detail = pd.read_csv(LOC / f"{NAME}_snps_detail.tsv", sep="\t")
snps = detail["varID"].tolist()


def dosages(pop):
    raw = pd.read_csv(LOC / f"{NAME}_{pop}.raw", sep="\t")
    x = raw.iloc[:, list(raw.columns).index("PHENOTYPE") + 1:]
    x.columns = [c.rsplit("_", 1)[0] for c in x.columns]
    return x[snps].to_numpy(dtype=float)


def ingredients(x):
    p = x.mean(axis=0) / 2
    d = 2 * p * (1 - p)
    with np.errstate(invalid="ignore", divide="ignore"):
        r = np.corrcoef(x, rowvar=False)
    r = np.nan_to_num(r)
    np.fill_diagonal(r, 1.0)
    return d, r


dE, RE = ingredients(dosages("EUR358"))
dY, RY = ingredients(dosages("YRI87"))
wE = detail["w_EUR"].to_numpy()
wY = detail["w_YRI"].to_numpy()


def logV(w, d, R):
    s = np.sqrt(d) * w
    return math.log(float(s @ R @ s))


players = ["w", "D", "R"]
vertex = {}
for bits in itertools.product([0, 1], repeat=3):
    w = wY if bits[0] else wE
    d = dY if bits[1] else dE
    R = RY if bits[2] else RE
    vertex["".join(map(str, bits))] = logV(w, d, R)

# Shapley value for each player: average marginal change over the 6 orders.
orders = list(itertools.permutations(range(3)))
paths = []
phi = np.zeros(3)
for order in orders:
    state = [0, 0, 0]
    steps = []
    for k in order:
        before = vertex["".join(map(str, state))]
        state[k] = 1
        after = vertex["".join(map(str, state))]
        steps.append({"player": players[k], "change": after - before})
        phi[k] += (after - before) / len(orders)
    paths.append({"order": [players[k] for k in order], "steps": steps})

dec = pd.read_csv(ROOT / "results" / "decomposition" / "EUR87_r1_vs_YRI87.tsv", sep="\t").set_index("gene").loc[GENE]
check = {"phi_w": abs(phi[0] - dec["phi_w"]), "phi_D": abs(phi[1] - dec["phi_D"]),
         "phi_R": abs(phi[2] - dec["phi_R"]), "delta": abs((vertex["111"] - vertex["000"]) - dec["delta"])}
worst = max(check.values())
if worst > 1e-6:
    raise SystemExit(f"animation numbers disagree with the decomposition output: {check}")

out = {
    "gene": NAME, "eur_set": "EUR87_r1", "yri_set": "YRI87", "n_snps": len(snps),
    "n_nonzero_eur": int((wE != 0).sum()), "n_nonzero_yri": int((wY != 0).sum()),
    "vertex_logV": vertex, "delta": vertex["111"] - vertex["000"],
    "phi": {"w": phi[0], "D": phi[1], "R": phi[2]}, "paths": paths,
    "weights_eur": wE.tolist(), "weights_yri": wY.tolist(),
    "max_abs_difference_vs_decomposition": worst,
}
(ROOT / "animation" / "dnajb7_numbers.json").write_text(json.dumps(out, indent=1))
print(f"8 vertices exported; delta {out['delta']:+.3f}; phi w {phi[0]:+.3f} D {phi[1]:+.3f} R {phi[2]:+.3f}; "
      f"max difference vs decomposition {worst:.2e}")
for p in paths:
    print(" -> ".join(f"{s['player']} {s['change']:+.3f}" for s in p["steps"]))

File: src/export_anim_numbers.R
## Export the real numbers the Manim animation shows, and prove they match the analysis.
## Refuses to write if any component disagrees with the decomposition by more than 1e-6.

suppressPackageStartupMessages({
  library(data.table)
  library(jsonlite)
})

find_root <- function() {
  a <- commandArgs(trailingOnly = FALSE)
  f <- sub("^--file=", "", a[grep("^--file=", a)])
  d <- if (length(f)) dirname(normalizePath(f[1L])) else getwd()
  normalizePath(file.path(d, ".."), mustWork = FALSE)
}
ROOT <- find_root()
LOC  <- file.path(ROOT, "results", "figdata", "locus")
GENE <- "ENSG00000172404.4"
NAME <- "DNAJB7"

detail <- fread(file.path(LOC, sprintf("%s_snps_detail.tsv", NAME)), sep = "\t")
snps   <- as.character(detail$varID)

dosages <- function(pop) {
  raw <- fread(file.path(LOC, sprintf("%s_%s.raw", NAME, pop)), sep = "\t")
  first <- which(names(raw) == "PHENOTYPE") + 1L
  x <- raw[, first:ncol(raw), with = FALSE]
  setnames(x, sub("_[^_]*$", "", names(x)))
  as.matrix(x[, snps, with = FALSE])
}

ingredients <- function(x) {
  p <- colMeans(x) / 2
  d <- 2 * p * (1 - p)
  r <- suppressWarnings(cor(x))
  r[!is.finite(r)] <- 0
  diag(r) <- 1
  list(d = d, R = r)
}

eur <- ingredients(dosages("EUR358"))
yri <- ingredients(dosages("YRI87"))
wE  <- as.numeric(detail$w_EUR)
wY  <- as.numeric(detail$w_YRI)

logV <- function(w, d, R) {
  s <- sqrt(d) * w
  log(as.numeric(t(s) %*% R %*% s))
}

players <- c("w", "D", "R")
bits_grid <- expand.grid(w = 0:1, D = 0:1, R = 0:1)
vertex <- list()
for (i in seq_len(nrow(bits_grid))) {
  b <- as.integer(bits_grid[i, ])
  w <- if (b[1L]) wY else wE
  d <- if (b[2L]) yri$d else eur$d
  R <- if (b[3L]) yri$R else eur$R
  vertex[[paste(b, collapse = "")]] <- logV(w, d, R)
}

## Shapley value for each player: average marginal change over the 6 orders.
orders <- list(c(1,2,3), c(1,3,2), c(2,1,3), c(2,3,1), c(3,1,2), c(3,2,1))
phi   <- numeric(3)
paths <- list()
for (o in orders) {
  state <- c(0L, 0L, 0L)
  steps <- list()
  for (k in o) {
    before <- vertex[[paste(state, collapse = "")]]
    state[k] <- 1L
    after <- vertex[[paste(state, collapse = "")]]
    steps[[length(steps) + 1L]] <- list(player = players[k], change = after - before)
    phi[k] <- phi[k] + (after - before) / length(orders)
  }
  paths[[length(paths) + 1L]] <- list(order = players[o], steps = steps)
}

dec <- fread(file.path(ROOT, "results", "decomposition", "EUR87_r1_vs_YRI87.tsv"), sep = "\t")
row <- dec[gene == GENE][1L]
check <- c(phi_w = abs(phi[1L] - row$phi_w), phi_D = abs(phi[2L] - row$phi_D),
           phi_R = abs(phi[3L] - row$phi_R),
           delta = abs((vertex[["111"]] - vertex[["000"]]) - row$delta))
worst <- max(check)
if (worst > 1e-6)
  stop(sprintf("animation numbers disagree with the decomposition output: %s",
               paste(sprintf("%s=%.2e", names(check), check), collapse = ", ")), call. = FALSE)

out <- list(gene = NAME, eur_set = "EUR87_r1", yri_set = "YRI87", n_snps = length(snps),
            n_nonzero_eur = sum(wE != 0), n_nonzero_yri = sum(wY != 0),
            vertex_logV = vertex, delta = vertex[["111"]] - vertex[["000"]],
            phi = list(w = phi[1L], D = phi[2L], R = phi[3L]), paths = paths,
            weights_eur = wE, weights_yri = wY,
            max_abs_difference_vs_decomposition = worst)
write(toJSON(out, auto_unbox = TRUE, digits = NA, pretty = TRUE),
      file.path(ROOT, "animation", "dnajb7_numbers.json"))

cat(sprintf("8 vertices exported; delta %+.3f; phi w %+.3f D %+.3f R %+.3f; max difference vs decomposition %.2e\n",
            out$delta, phi[1L], phi[2L], phi[3L], worst))
for (p in paths)
  cat(paste(vapply(p$steps, function(s) sprintf("%s %+.3f", s$player, s$change), character(1L)),
            collapse = " -> "), "\n")
